### PR TITLE
refactor: remove async new

### PR DESCRIPTION
**Pull request by ZENOTME**
I find that most of the `async new` due to the `async new` in rolling writer. We can avoid them by delay creating inner writer in rolling writer.

### CODE REVIEW

File: icelake/src/io/appender/rolling_writer.rs [hunk 1]
```diff
@@ -163,6 +167,12 @@ impl RollingWriter {
 impl FileAppender for RollingWriter {
     /// Write a record batch. The `DataFileWriter` will create a new file when the current row num is greater than `target_file_row_num`.
     async fn write(&mut self, batch: RecordBatch) -> Result<()> {
+        // # TODO
+        // Optimize by `unlikely`
```
Comment (ZENOTME) [inline icelake/src/io/appender/rolling_writer.rs:171]: Seems unlikely still nightly-only now. https://doc.rust-lang.org/std/intrinsics/fn.unlikely.html

**Comment by liurenjie1024** [comment]
What's the motivation behind this? Why we want to remove `async` in `new` method? TBH I'm in favor of current approach since we have enough type safety without adding unnecessary `None` check.

**Comment by ZENOTME** [comment]
> What's the motivation behind this? Why we want to remove async in new method? TBH I'm in favor of current approach since we have enough type safety without adding unnecessary None check.

I find that if we remove the asyn new in rolling writer so that lots of new can be sync including the build function in `RecordBatchBuilder`. The benefit is a sync function may more general then a async. But I find that it has to introduce `None` check. So it's ok not to merge it, I just trying to see whether it's worth it.🤣

**Comment by liurenjie1024** [comment]
> > What's the motivation behind this? Why we want to remove async in new method? TBH I'm in favor of current approach since we have enough type safety without adding unnecessary None check.
> 
> I find that if we remove the asyn new in rolling writer so that lots of new can be sync including the build function in `RecordBatchBuilder`. The benefit is a sync function may more general then a async. But I find that it has to introduce `None` check. So it's ok not to merge it, I just trying to see whether it's worth it.🤣

The two color function problem is a a problem of async rust. Personally I prefer type safe rather than async. cc @Xuanwo What do you think?

**Comment by ZENOTME** [comment]
> Personally I prefer type safe rather than async.

How about wrapping this Inner File Writer so that the type can be used in safe.
```
struct InnerFileWriter {
  operator: Operator,
  parquet_config: ..,
  location: String,
  row_num: usize,
  writer: Option<Writer>
}

impl InnerFileWriter {
  pub new(location: &str) -> Self;
  pub row_num(&self) -> usize;
  async pub write(&self, batch) -> {
    // guarantee to init the writer if it's None.
  }
}
```

**Comment by liurenjie1024** [comment]
> > Personally I prefer type safe rather than async.
> 
> How about wrapping this Inner File Writer so that the type can be used in safe.
> 
> ```
> struct InnerFileWriter {
>   operator: Operator,
>   parquet_config: ..,
>   location: String,
>   row_num: usize,
>   writer: Option<Writer>
> }
> 
> impl InnerFileWriter {
>   pub new(location: &str) -> Self;
>   pub row_num(&self) -> usize;
>   async pub write(&self, batch) -> {
>     // guarantee to init the writer if it's None.
>   }
> }
> ```

Sounds reasonable to me. Instead of using `Option`, is it possible to use sth like `Once`?

**Comment by Xuanwo** [comment]
I believe this location isn't a critical path where most IO occurs on the writer rather than an `async new`. I attempted something similar in OpenDAL, where I tried to remove some `async` functions. However, I discovered that:

- No observable impact on performance.
- Complexity of the code has significantly improved.
- No one really use them: they always calling `Xxx::new` under async context.

**Comment by ZENOTME** [comment]
> I believe this location isn't a critical path where most IO occurs on the writer rather than an `async new`. I attempted something similar in OpenDAL, where I tried to remove some `async` functions. However, I discovered that:
> 
> * No observable impact on performance.
> * Complexity of the code has significantly improved.
> * No one really use them: they always calling `Xxx::new` under async context.

Thanks! It's useful! Indeed, complexity of the code has significantly improved and seems the benefit is not also observable. Let's close this PR🥵.

**Pull request closed without merging**